sql: fix the handling of integer types

knz opened the pull request:
Addresses a large chunk of #26925.
Fixes #25098.
Informs #24686.

Prior to this patch, CockroachDB maintained an unnecessary distinction
between "INT" and "INTEGER", between "BIGINT" and "INT8", etc.

This distinction is unnecessary but also costly, as we were paying the
price of a "name" attribute in coltypes.TInt, with a string comparison
and hash table lookup on every use of the type.

What really matters is that the type shows up properly in
introspection; this has already been ensured by various
OID-to-pgcatalog mappings and the recently introduced
`InformationSchemaTypeName()`.

Any distinction beyond that is unnecessary and can be dropped from the
implementation.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28690)
<!-- Reviewable:end -->

knz commented:
The PR  is not yet ready for review. Will ping.

knz commented:
ok I think the review can start now.

What's missing:

- fixing some more test
- introduce a SQL migration to populate the missing Width field on FLOAT column descriptors.

BramGruneir reviewed (dismissed):
> Reviewed 39 of 44 files at r1, 20 of 20 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/create_table.go, line 1405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2B7bH2vTZECrQ4XRs:-LK2B7bH2vTZECrQ4XRt:bsql39u) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/create_table.go#L1405)):*
> ```Go
> 	evalCtx *tree.EvalContext,
> ) error {
> 	if d.HasDefaultExpr() {
> ```

This code movement seems unrelated to this PR, but this check should indeed come at the start.
___
*[pkg/sql/information_schema.go, line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2BPEk6fruLn7O7uNw:-LK2BPEk6fruLn7O7uNx:b-uazq8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/information_schema.go#L288)):*
> ```Go
> 					tree.DNull,                                          // character_set_name
> 					dStringPtrOrEmpty(column.ComputeExpr),               // generation_expression
> 					tree.NewDString(column.Type.SQLString()),            // crdb_sql_type
> ```

This change needs to be addressed in docs, and should probably be mentioned in the release notes.
___
*[pkg/sql/show_columns.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2SzMC7VTTIq40UPER:-LK2SzMD5VeBogtDNM_Y:bj9alyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/show_columns.go#L37)):*
> ```Go
>   IF(inames[1] IS NULL, ARRAY[]:::STRING[], inames) AS indices
> FROM
>   (SELECT column_name, crdb_sql_type, data_type, is_nullable, column_default, generation_expression, ordinal_position,
> ```

is data type used at all anymore?
___
*[pkg/sql/coltypes/aliases.go, line 143 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2CwH679I84jsSU58O:-LK2CwH679I84jsSU58P:byn2rbv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/coltypes/aliases.go#L143)):*
> ```Go
> // NewFloat creates a type alias for FLOAT with the given precision.
> func NewFloat(prec int64) (*TFloat, error) {
> 	if prec < 1 {
> ```

just for style, perhaps a switch here would be a bit cleaner
___
*[pkg/sql/coltypes/arith.go, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2E7KzEsCyW_rPxXCq:-LK2E7KzEsCyW_rPxXCr:bhl52iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/coltypes/arith.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> 
> package coltypes
> ```

This is just me personally, but I would really like to see each type get it's own file.  This goes for all types.  Maybe a future PR.
___
*[pkg/sql/coltypes/arith.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2ENFv6K0Taab9b1Ee:-LK2ENFv6K0Taab9b1Ef:b2gjxb6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/coltypes/arith.go#L30)):*
> ```Go
> func (node *TBool) TypeName() string { return "BOOL" }
> 
> // PGTypeName implements the ColTypeFormatter interface.
> ```

I'm a fan of ensuring an interface is implemented using the dummy declaration.  Just makes me feel safer.
var _ foo.RequiredInterface = myType{}
___
*[pkg/sql/logictest/testdata/planner_test/distsql_agg, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2fkyFEzS5QmYJE1Gm:-LK2fkyFEzS5QmYJE1Gn:bwghzg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/logictest/testdata/planner_test/distsql_agg#L126)):*
> ```
> SELECT url FROM [EXPLAIN (DISTSQL) SELECT avg(c), sum(c), avg(d), sum(d) FROM data]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html#eJy8lF-L4jAUxd_3U8h9WiFg80fH6VNkYEHYGRf_PC1lyZpLEZxGkhR2GfzuS1thWrFpwbpPmtuc_E7ODfcDMqPxTb2jg_gnUCDAgAAHAgIITCEhcLJmj84ZW2ypBEv9B-KIwCE75b4oJwT2xiLEH-AP_ogQw1b9PuIalUY7iYCARq8OxxJzsod3Zf9KrbwCAqvcxyPJiRSQnAmY3H-e6rxKEWJ6Jv3JizS1mCpv7GTaBG92r18lHQOBl9XubXv5X1ZZrcrGrUZYq5FPfp4Zq9GibsCTc9gqjdq8bnavv5YXXxe3vFEXxWqNmUYbjySdvCw2xe7F5tv31WI7H5ORpGQk-USK4qf1crxxOdq_v3TY_naQa6HNHttf1j8CNmwEHeRaBE-PjYD3j4APG0EHuRbB_LERiP4RiGEj6CDXInj-f4PuhpE1upPJHF4NvNsnR8UgRJ1iNTWdye0ef1izLzHVclXqyoJG56uvtFoss-pTYbAupkExa4jptZiFyR1oHlSLsFjc43saFM_C5Nk95KegeB4mz-8hP4d7FXU8k_Aju2Yn5y__AgAA___R7uSX
> ```

Why did these change?
___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_agg, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2fugB7-01j3RJHABJ:-LK2fugB7-01j3RJHABK:bwghzg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/opt/exec/execbuilder/testdata/distsql_agg#L126)):*
> ```
> SELECT url FROM [EXPLAIN (DISTSQL) SELECT avg(c), sum(c), avg(d), sum(d) FROM data]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html#eJy8lF-L4jAUxd_3U8h9WiFg80fH6VNkYEHYGRf_PC1lyZpLEZxGkhR2GfzuS1thWrFpwbpPmtuc_E7ODfcDMqPxTb2jg_gnUCDAgAAHAgIITCEhcLJmj84ZW2ypBEv9B-KIwCE75b4oJwT2xiLEH-AP_ogQw1b9PuIalUY7iYCARq8OxxJzsod3Zf9KrbwCAqvcxyPJiRSQnAmY3H-e6rxKEWJ6Jv3JizS1mCpv7GTaBG92r18lHQOBl9XubXv5X1ZZrcrGrUZYq5FPfp4Zq9GibsCTc9gqjdq8bnavv5YXXxe3vFEXxWqNmUYbjySdvCw2xe7F5tv31WI7H5ORpGQk-USK4qf1crxxOdq_v3TY_naQa6HNHttf1j8CNmwEHeRaBE-PjYD3j4APG0EHuRbB_LERiP4RiGEj6CDXInj-f4PuhpE1upPJHF4NvNsnR8UgRJ1iNTWdye0ef1izLzHVclXqyoJG56uvtFoss-pTYbAupkExa4jptZiFyR1oHlSLsFjc43saFM_C5Nk95KegeB4mz-8hP4d7FXU8k_Aju2Yn5y__AgAA___R7uSX
> ```

Why did these change?
___
*[pkg/sql/opt/memo/private_storage_test.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2gCUa4eOGfavYSS-g:-LK2gCUbA_Lp1Zj3WUkF:blv4ro8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/opt/memo/private_storage_test.go#L373)):*
> ```Go
> 	nf30, _ := coltypes.NewFloat(16)
> 	test(coltypes.Float, nf30, false)
> 	test(nf30, nf30, true)
> ```

What is this line testing?
___
*[pkg/sql/parser/parse_test.go, line 1139 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2jBsU2G0TlvCXfENh:-LK2jBsU2G0TlvCXfENi:b-6e0qve) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/parser/parse_test.go#L1139)):*
> ```Go
> 		{`SELECT NUMERIC 'foo'`, `SELECT DECIMAL 'foo'`},
> 		{`SELECT REAL 'foo'`, `SELECT FLOAT4 'foo'`},
> 		{`SELECT DOUBLE PRECISION 'foo'`, `SELECT FLOAT8 'foo'`},
> ```

Maybe add an array or two as well?
___
*[pkg/sql/parser/sql.y, line 5895 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2ke_e65uurJ32aJSs:-LK2ke_e65uurJ32aJSt:bdf35xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/parser/sql.y#L5895)):*
> ```Yacc
> | const_datetime
> | const_json
> | BLOB
> ```

There's no way to just collapse all BLOB, BYTES and BYTEA into one is there?  It would just look so much nicer.
___
*[pkg/sql/sem/tree/col_types_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2gktkEga9DzorqUyj:-LK2gktkEga9DzorqUyk:b88377g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/sem/tree/col_types_test.go#L30)):*
> ```Go
> 	testData := []struct {
> 		str          string
> 		norm         string
> ```

please add a comment about this being empty means the name matches.
___
*[pkg/sql/sem/tree/parse_array.go, line 151 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2hEUG0GE2BMICTFCt:-LK2hEUG0GE2BMICTFCu:bo73q9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/sem/tree/parse_array.go#L151)):*
> ```Go
> // string representation of the type. Used by dump. It only
> // supports those type names that can appear immediately before `[]`.
> func ArrayElementTypeStringToColType(s string) (coltypes.T, error) {
> ```

I think this should be a function in coltypes.  Ideally inside arrays.go. 
___
*[pkg/sql/sem/tree/testdata/pretty/7.ref.golden, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2mG6Z4Z09bLjnITBq:-LK2mG6Z4Z09bLjnITBr:by1ma4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/56c92496d678db571e756e1be412fa7ab60518d9/pkg/sql/sem/tree/testdata/pretty/7.ref.golden#L136)):*
> ```
> 	NOTHING
> 
> 26:
> ```

Why did you remove this test case?
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 26 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK33f6T-2xCdGbH4fDS:-LK33f6T-2xCdGbH4fDT:b-56udd2) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L26)):*
> ```Go
> 	"github.com/pkg/errors"
> )
> ```

All of the functions in this file should have unit tests.
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK32_Cj9xlIJFPyUUrf:-LK32_Ck0m5-a5zU1DVM:b-6ivbn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L31)):*
> ```Go
> // After calling this, the caller must also call PopulateTypeAttrs
> // below.
> func DatumTypeToColumnType(ptyp types.T) (ColumnType, error) {
> ```

This has a mix of some returns inside the cases and some out.

Can you move them all in and get rid of the ctyp variable entirely.
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 92 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK33LSZ4hbySpR_99Lv:-LK33LSZ4hbySpR_99Lw:b-ye3jkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L92)):*
> ```Go
> 			case 16:
> 				base.VisibleType = ColumnType_SMALLINT
> 			case 32:
> ```

these two lying cases can be joined together.
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK33xN-CK_O03lr_bAd:-LK33xN-CK_O03lr_bAe:b-kvncj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L180)):*
> ```Go
> 
> // InfoSchemaColumnType returns the string suitable to populate the data_type column
> // of information_schema.columns.
> ```

I'm confused about information_schema.columns.

Are postgres type names compatible but we're not?  Because these look like postgres type names.

For pg_catalog, let's go with postgres' type names, but for information schema, let's either go with our type names unless there's another standard we should match.
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 279 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK34hIgB5RunsvpULe_:-LK34hIgB5RunsvpULea:bislota) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L279)):*
> ```Go
> 			// Pre-2.1 columns: the width is not set yet and instead there
> 			// is a precision. Reverse-engineer the width from that.
> 			if c.Precision < 0 {
> ```

switch here might make it a lot nicer to read

Also, this matches  FloatProperties(), any way to DRY it up?  Especially considering that this has some extra strange logic in it.
___
*[pkg/sql/sqlbase/structured.go, line 2221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2nKhD3RnPXOLnRvzQ:-LK2nKhD3RnPXOLnRvzR:b-sd3d7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.go#L2221)):*
> ```Go
> 		} else if c.Precision == 0 {
> 			// Special case of poorly initialized coltypes pre-2.1.
> 			width = 64
> ```

why not just return here?  And for all the cases?  It will greatly simplify the logic:

```go
func (c *ColumnType) FloatProperties() (int32, int32) {
  if c.width == 0 {
    // Pre-2.1 columns: the width is not set yet and instead there
    // is a precision. Reverse-engineer the width from that.
    if c.Precision < 0 {
      panic(fmt.Sprintf("programming error: invalid float precision: %d", c.Precision))
    }
    if c.Precision == 0 {
      // Special case of poorly initialized coltypes pre-2.1.
      return int32(64), int32(53)
     }
    if c.Precision <= 24 {
      return int32(32), int32(24)
    } 
    if c.Precision <= 54 {
      return int32(64), int32(53)
    }
    panic(fmt.Sprintf("programming error: invalid float precision: %d", c.Precision))
  }

  if c.width == 64 {
    return int32(64), int32(53)
  }
  return c.width, int32(24)
}
```

might even be worth adding a switch into the inner if
___
*[pkg/sql/sqlbase/structured.go, line 2240 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2pv353eX3IIEw2Lyg:-LK2pv363l-83YmnOt4k:b-9i6fk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.go#L2240)):*
> ```Go
> // data types. Returns false if the data type is not numeric, or if the precision
> // of the numeric type is not bounded.
> func (c *ColumnType) NumericPrecision() (int32, bool) {
> ```

Does this ever return true for 0?  If not, why not remove the bool return value entirely?
___
*[pkg/sql/sqlbase/structured.proto, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK35h9Z7GOiPc1Et0CF:-LK35h9Z7GOiPc1Et0CG:b-4s3qz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.proto#L68)):*
> ```protobuf
>   // Note: this enum is becoming deprecated. Do not add items
>   // to it without consideration.
>   // TODO(knz): remove this.
> ```

Please add to this comment as to when you can do so. 
___
*[pkg/sql/sqlbase/structured.proto, line 80 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2qN7Q1Wom5gALwSjA:-LK2qN7Q1Wom5gALwSjB:bgjw5cc) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.proto#L80)):*
> ```protobuf
> 
>   optional SemanticType semantic_type = 1 [(gogoproto.nullable) = false];
>   // BIT, INT, FLOAT, DECIMAL, CHAR, BINARY, FLOAT.
> ```

nit: this isn't a sentence, no period at the end
___
*[pkg/sql/sqlbase/structured.proto, line 82 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2qQnQ7kF0So-Q619H:-LK2qQnQ7kF0So-Q619I:bqrvhhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.proto#L82)):*
> ```protobuf
>   // BIT, INT, FLOAT, DECIMAL, CHAR, BINARY, FLOAT.
>   optional int32 width = 2 [(gogoproto.nullable) = false];
>   // DECIMAL.
> ```

nit: this isn't a sentence, no period at the end

Perhaps 
// DECIMAL, pre2.1 FLOAT (incorrectly)


<!-- Sent from Reviewable.io -->

BramGruneir reviewed (commented):
We also need to ensure that docs is aware of the changes and I think the release note need to be greatly expanded.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-:-LK37f7T1Av3Sb11boXP:b-tdnkjj)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
Thank you for your initial review. I'll iterate on that today.

Regarding the doc work I think I will simply take care of doing all the doc updates myself. There is no way the doc team will have the bandwidth to process such a change themselves.

In the meantime you can have a look at the SERIAL PR #28575

knz reviewed (commented):
Some reactions already, will process the rest tomorrow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-:-LK4_xHJ9jSURRpF1rGq:b-kgyk90)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/show_columns.go, line 37 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2SzMC7VTTIq40UPER:-LK4ZV_W4bH_L29osKOk:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/show_columns.go#L37)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

is data type used at all anymore?
</blockquote></details>

Removed.
___
*[pkg/sql/coltypes/arith.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2ENFv6K0Taab9b1Ee:-LK4ZbPAAljszXkBCGf0:bb5mdsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/coltypes/arith.go#L30)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I'm a fan of ensuring an interface is implemented using the dummy declaration.  Just makes me feel safer.
var _ foo.RequiredInterface = myType{}
</blockquote></details>

This is already done in `interface.go`.
___
*[pkg/sql/logictest/testdata/planner_test/distsql_agg, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2fkyFEzS5QmYJE1Gm:-LK4ZfBL6uJ1jAoCKC7P:biiy8n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/logictest/testdata/planner_test/distsql_agg#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why did these change?
</blockquote></details>

FLOAT -> FLOAT8
___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_agg, line 126 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2fugB7-01j3RJHABJ:-LK4ZhOx0PjqACsnnzli:biiy8n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/opt/exec/execbuilder/testdata/distsql_agg#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why did these change?
</blockquote></details>

FLOAT -> FLOAT8
___
*[pkg/sql/opt/memo/private_storage_test.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2gCUa4eOGfavYSS-g:-LK4ZoQQCIOELRzdyGsg:b10o642) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/opt/memo/private_storage_test.go#L373)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

What is this line testing?
</blockquote></details>

That interning two times the same type yields equal instances. I had botched the test, it is now fixed.
___
*[pkg/sql/sem/tree/testdata/pretty/7.ref.golden, line 136 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2mG6Z4Z09bLjnITBq:-LK4_Gy1Cg0V-3KCLP7z:bpb7m4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/56c92496d678db571e756e1be412fa7ab60518d9/pkg/sql/sem/tree/testdata/pretty/7.ref.golden#L136)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Why did you remove this test case?
</blockquote></details>

This is auto-generated. The test case disappears because there is no difference in output for this column width.
___
*[pkg/sql/sqlbase/coltype_conversion.go, line 180 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK33xN-CK_O03lr_bAd:-LK4_PUT5FiqD6LV7cYv:b-75cf1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/coltype_conversion.go#L180)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I'm confused about information_schema.columns.

Are postgres type names compatible but we're not?  Because these look like postgres type names.

For pg_catalog, let's go with postgres' type names, but for information schema, let's either go with our type names unless there's another standard we should match.
</blockquote></details>

Yes I was confused myself. ACtually there are two completely unrelated things: 

- the "pg display name", which is the internal name of the type inside postgresql  and which we use in pg_catalog.
- the information_schema type, which is specified by the SQL standard, and which is generated here.
___
*[pkg/sql/sqlbase/structured.go, line 2221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2nKhD3RnPXOLnRvzQ:-LK4_ipo1fM_16ZydqBL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.go#L2221)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

why not just return here?  And for all the cases?  It will greatly simplify the logic:

```go
func (c *ColumnType) FloatProperties() (int32, int32) {
  if c.width == 0 {
    // Pre-2.1 columns: the width is not set yet and instead there
    // is a precision. Reverse-engineer the width from that.
    if c.Precision < 0 {
      panic(fmt.Sprintf("programming error: invalid float precision: %d", c.Precision))
    }
    if c.Precision == 0 {
      // Special case of poorly initialized coltypes pre-2.1.
      return int32(64), int32(53)
     }
    if c.Precision <= 24 {
      return int32(32), int32(24)
    } 
    if c.Precision <= 54 {
      return int32(64), int32(53)
    }
    panic(fmt.Sprintf("programming error: invalid float precision: %d", c.Precision))
  }

  if c.width == 64 {
    return int32(64), int32(53)
  }
  return c.width, int32(24)
}
```

might even be worth adding a switch into the inner if
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.proto, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK35h9Z7GOiPc1Et0CF:-LK4_ozfCQXC-8lxRStA:b-4nab0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/98a3278aeb59ed4350cb9d9de9f974c18907ef97/pkg/sql/sqlbase/structured.proto#L68)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Please add to this comment as to when you can do so. 
</blockquote></details>

I was mistaken; instead I reverted that change and went in a different direction.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/create_table.go, line 1405 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-LK2B7bH2vTZECrQ4XRs:-LK6tBHU59T9McnBAM4h:b-x1x4aa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7dd8b4b81bb1332fc5d4ab35d968b9e848cca81/pkg/sql/create_table.go#L1405)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

This code movement seems unrelated to this PR, but this check should indeed come at the start.
</blockquote></details>

Forked as #28753.


<!-- Sent from Reviewable.io -->

knz committed:
sql: fix the handling of integer types

Prior to this patch, CockroachDB maintained an unnecessary distinction
between "INT" and "INTEGER", between "BIGINT" and "INT8", etc.

This distinction is unnecessary but also costly, as we were paying the
price of a "name" attribute in coltypes.TInt, with a string comparison
and hash table lookup on every use of the type.

What really matters is that the type shows up properly in
introspection; this has already been ensured by various
OID-to-pgcatalog mappings and the recently introduced
`InformationSchemaTypeName()`.

Any distinction beyond that is unnecessary and can be dropped from the
implementation.

Release note: None

knz commented:
And we're done! All the bits have been forked off, and the remaining PR (the last commit in this series) is now minuscule.

Also ready for review.

knz commented:
rebased, RFAL

BramGruneir reviewed (commented):
@knz, can you pull this out into a fresh PR?  I'm having some issues trying to review only the final commit.

> Reviewed 8 of 56 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-:-LKcLd5yDB4YTiye4LjC:b-e16q3c)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> can you pull this out into a fresh PR?

My pleasure: #29020.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28690#-:-LKcbxZREKctxcj3AnuA:bg8d4w)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
bors r+

BramGruneir commented:
I think you're borsing the wrong PR now.  Unless you updated the commit in this PR as well.

BramGruneir commented:
But probably best to close this one and merge the other pr.

knz commented:
They have the same commit SHAs, just different PR numbers. You'll see, when one merges the other will be merged automatically at the same time.
